### PR TITLE
Bug 1712072 - Update distribution docs to new structure

### PR DESCRIPTION
**Pull request by badboy**

### CODE REVIEW

File: docs/user/reference/metrics/memory_distribution.md [hunk 1]
```diff
@@ -9,269 +9,392 @@ That is, the function from a value \\( x \\) to a bucket index is:
...
+fn allocate_memory(bytes: u64) {
     // ...
-    Memory.heapAllocated.accumulate(nbytes / 1024)
+    memory::heap_allocated.accumulate(bytes / 1024);
 }
 ```
 
-There are test APIs available too.  For convenience, properties `sum` and `count` are exposed to facilitate validating that data was recorded correctly.
+</div>
+<div data-lang="JavaScript" class="tab"></div>
```
Comment (brizental) [inline docs/user/reference/metrics/memory_distribution.md:82]:
```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716952"></div>
```

File: docs/user/reference/metrics/memory_distribution.md [hunk 2]
```diff
@@ -9,269 +9,392 @@ That is, the function from a value \\( x \\) to a bucket index is:
...
+Glean.memory.heapAllocated.accumulate(bytes / 1024);
+```
+
+</div>
+
+{{#include ../../../shared/tab_footer.md}}
+
+#### Recorded errors
+
+* `invalid_value`: If recording a negative memory size.
```
Comment (brizental) [inline docs/user/reference/metrics/memory_distribution.md:105]:
```suggestion
* [`invalid_value`](../../user/metrics/error-reporting.md): If recording a negative memory size.
```

Same below, or merge the two bullet points into one.

File: docs/user/reference/metrics/memory_distribution.md [hunk 3]
```diff
@@ -9,269 +9,392 @@ That is, the function from a value \\( x \\) to a bucket index is:
...
+* `invalid_value`: If recording a size larger than 1 TB.
+
+## Testing API
+
+### `testGetValue`
+
+Gets the recorded value for a given counter metric.
```
Comment (brizental) [inline docs/user/reference/metrics/memory_distribution.md:112]:
```suggestion
Gets the recorded value for a given memory distribution metric.
```

File: docs/user/reference/metrics/memory_distribution.md [hunk 4]
```diff
@@ -9,269 +9,392 @@ That is, the function from a value \\( x \\) to a bucket index is:
...
 </div>
+<div data-lang="JavaScript" class="tab"></div>
+<div data-lang="Firefox Desktop" class="tab"></div>
 
 {{#include ../../../shared/tab_footer.md}}
 
-## Limits
+## Metric parameters
+
+Example custom distribution metric definition:
```
Comment (brizental) [inline docs/user/reference/metrics/memory_distribution.md:359]:
```suggestion
Example memory distribution metric definition:
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 5]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
 
-// Assert that no errors were recorded.
-assertEquals(0, Pages.pageLoad.testGetNumRecordedErrors(ErrorType.InvalidValue))
+fn on_page_start() {
+    self.timer_id = pages::page_load.start();
+}
 ```
 
 </div>
+<div data-lang="JavaScript" class="tab"></div>
```
Comment (brizental) [inline docs/user/reference/metrics/timing_distribution.md:114]:
```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 6]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
+
+</div>
+
+{{#include ../../../shared/tab_footer.md}}
+
+#### Recorded errors
+
+* `invalid_value`: If recording a negative timespan.
+* `invalid_state`: If a non-existing/stopped timer is stopped again.
+* `invalid_overflow`: If recording a time longer than the maximum for the given unit.
```
Comment (brizental) [inline docs/user/reference/metrics/timing_distribution.md:225]: Linkify as above.

File: docs/user/reference/metrics/timing_distribution.md [hunk 7]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
+
+fn on_page_loaded() {
+    pages::page_load.stop_and_accumulate(self.timer_id);
 }
+```
 
-func onPageLoaded() {
-    Pages.pageLoad.stopAndAccumulate(timerId)
+</div>
+<div data-lang="JavaScript" class="tab"></div>
```
Comment (brizental) [inline docs/user/reference/metrics/timing_distribution.md:200]:
```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 8]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
+
+</div>
+
+{{#include ../../../shared/tab_footer.md}}
+
+## Testing API
+
+### `testGetValue`
+
+Gets the recorded value for a given counter metric.
```
Comment (brizental) [inline docs/user/reference/metrics/timing_distribution.md:361]:
```suggestion
Gets the recorded value for a given timing distribution metric.
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 9]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
+<div data-lang="Firefox Desktop" class="tab"></div>
 
-auto timerId = mozilla::glean::pages::page_load.Start();
-PR_Sleep(PR_MillisecondsToInterval(10));
-mozilla::glean::pages::page_load.StopAndAccumulate(std::move(timerId));
+{{#include ../../../shared/tab_footer.md}}
+
+### `testGetNumRecordedErrors`
+
+Gets number of errors recorded for a given counter metric.
```
Comment (brizental) [inline docs/user/reference/metrics/timing_distribution.md:517]:
```suggestion
Gets number of errors recorded for a given timing distribution metric.
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 10]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
 
 </div>
+<div data-lang="JavaScript" class="tab"></div>
+<div data-lang="Firefox Desktop" class="tab"></div>
 
 {{#include ../../../shared/tab_footer.md}}
 
+## Metric parameters
+
+Example custom distribution metric definition:
```
Comment (brizental) [inline docs/user/reference/metrics/timing_distribution.md:589]:
```suggestion
Example timing distribution metric definition:
```

File: docs/user/reference/metrics/memory_distribution.md [hunk 11]
```diff
@@ -9,269 +9,392 @@ That is, the function from a value \\( x \\) to a bucket index is:
...
+assert_eq!(11, snapshot.sum);
+
+// Usually you don't know the exact timing values,
+// but how many should have been recorded.
+assert_eq!(2, snapshot.values.len());
 ```
 
+
 </div>
+<div data-lang="JavaScript" class="tab"></div>
```
Comment (brizental) [inline docs/user/reference/metrics/memory_distribution.md:204]:
```suggestion
<div data-lang="JavaScript" class="tab"  data-bug="1716952"></div>
```

File: docs/user/reference/metrics/memory_distribution.md [hunk 12]
```diff
@@ -9,269 +9,392 @@ That is, the function from a value \\( x \\) to a bucket index is:
...
+use glean_metrics::pages;
+
+assert_eq!(
+    0,
+    pages::page_load.test_get_num_recorded_errors(ErrorType::InvalidValue)
+);
 ```
 
 </div>
+<div data-lang="JavaScript" class="tab"></div>
```
Comment (brizental) [inline docs/user/reference/metrics/memory_distribution.md:352]:
```suggestion
<div data-lang="JavaScript" class="tab"  data-bug="1716952"></div>
```

File: docs/user/reference/metrics/memory_distribution.md [hunk 13]
```diff
@@ -9,269 +9,392 @@ That is, the function from a value \\( x \\) to a bucket index is:
...
-Assert.True(Memory.heapAllocated.TestHasValue());
+{{#include ../../../shared/tab_footer.md}}
 
-// Get snapshot
-var snapshot = Memory.heapAllocated.TestGetValue();
+### `testHasValue`
 
-// Does the sum have the expected value?
-Assert.Equal(11, snapshot.Sum);
+Whether or not **any** value was recorded for a given counter metric.
```
Comment (brizental) [inline docs/user/reference/metrics/memory_distribution.md:230]:
```suggestion
Whether or not **any** value was recorded for a given memory distribution metric.
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 14]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
+</div>
+
+{{#include ../../../shared/tab_footer.md}}
 
-Continuing the `pageLoad` example above, at this point the metric should have a `sum == 11` and a `count == 2`:
+### `testHasValue`
 
-```C#
-using static Mozilla.YourApplication.GleanMetrics.Pages;
+Whether or not **any** value was recorded for a given counter metric.
```
Comment (brizental) [inline docs/user/reference/metrics/timing_distribution.md:463]:
```suggestion
Whether or not **any** value was recorded for a given timing distribution metric.
```

File: docs/user/reference/metrics/memory_distribution.md [hunk 15]
```diff
@@ -9,269 +9,392 @@ That is, the function from a value \\( x \\) to a bucket index is:
...
 
-<div data-lang="C++" class="tab">
+{{#include ../../../shared/tab_footer.md}}
 
-> **Note**: C++ APIs are only available in Firefox Desktop.
+### `testGetNumRecordedErrors`
 
-```c++
-#include "mozilla/glean/GleanMetrics.h"
+Gets number of errors recorded for a given counter metric.
```
Comment (brizental) [inline docs/user/reference/metrics/memory_distribution.md:284]:
```suggestion
Gets number of errors recorded for a given memory distribution metric.
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 16]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
-Continuing the `pageLoad` example above, at this point the metric should have a `sum == 11` and a `count == 2`:
+with metrics.pages.page_load.measure():
+    # Load a page
+```
 
-```Swift
-@testable import Glean
+</div>
+<div data-lang="Rust" class="tab"></div>
+<div data-lang="JavaScript" class="tab"></div>
```
Comment (brizental) [inline docs/user/reference/metrics/timing_distribution.md:268]:
```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 17]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
+let snapshot = pages::page_load.test_get_value(None).unwrap();
+
+// Usually you don't know the exact timing values,
+// but how many should have been recorded.
+assert_eq!(1, snapshot.values.len());
 ```
 
+
 </div>
+<div data-lang="JavaScript" class="tab"></div>
```
Comment (brizental) [inline docs/user/reference/metrics/timing_distribution.md:438]:
```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 18]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
+use glean_metrics::pages;
+
+assert_eq!(
+    0,
+    pages::page_load.test_get_num_recorded_errors(ErrorType::InvalidValue)
+);
 ```
 
 </div>
+<div data-lang="JavaScript" class="tab"></div>
```
Comment (brizental) [inline docs/user/reference/metrics/timing_distribution.md:582]:
```suggestion
 <div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

**Review by brizental — APPROVED** [review]

**Comment by brizental** [comment]
This is missing custom distribution. Would you mind filing a follow up for that?

Otherwise looks good. Thanks for taking it.

**Comment by badboy** [comment]
> This is missing custom distribution. Would you mind filing a follow up for that?
> 
> Otherwise looks good. Thanks for taking it.

[Custom Distribution has been migrated already, mostly](https://mozilla.github.io/glean/book/reference/metrics/custom_distribution.html)

File: docs/user/reference/metrics/memory_distribution.md [hunk 19]
```diff
@@ -9,269 +9,392 @@ That is, the function from a value \\( x \\) to a bucket index is:
...
+```
+
+</div>
+
+{{#include ../../../shared/tab_footer.md}}
+
+#### Recorded errors
+
+* `invalid_value`: If recording a negative memory size.
+* `invalid_value`: If recording a size larger than 1 TB.
```
Comment (badboy) [inline docs/user/reference/metrics/memory_distribution.md:106]:
```suggestion
* [`invalid_value`](../../user/metrics/error-reporting.md): If recording a size larger than 1 TB.
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 20]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
+Glean.pages.pageLoad.stopAndAccumulate(timerId);
+```
+
+</div>
+
+{{#include ../../../shared/tab_footer.md}}
+
+#### Recorded errors
+
+* `invalid_value`: If recording a negative timespan.
```
Comment (badboy) [inline docs/user/reference/metrics/timing_distribution.md:223]:
```suggestion
* [`invalid_value`](../../user/metrics/error-reporting.md): If recording a negative timespan.
```

File: docs/user/reference/metrics/timing_distribution.md [hunk 21]
```diff
@@ -63,316 +55,584 @@ val timerId : GleanTimerId
...
+* `invalid_state`: If a non-existing/stopped timer is stopped again.
```
Comment (badboy) [inline docs/user/reference/metrics/timing_distribution.md:224]:
```suggestion
* [`invalid_state`](../../user/metrics/error-reporting.md): If a non-existing/stopped timer is stopped again.
```

File: docs/user/reference/metrics/timing_distribution.md (hunk 6 above)
Comment (badboy) [reply docs/user/reference/metrics/timing_distribution.md:225]:
```suggestion
* [`invalid_overflow`](../../user/metrics/error-reporting.md): If recording a time longer than the maximum for the given unit.
```

File: docs/user/reference/metrics/timing_distribution.md (hunk 18 above)
Comment (badboy) [reply docs/user/reference/metrics/timing_distribution.md:582]:
```suggestion
<div data-lang="JavaScript" class="tab" data-bug="1716954"></div>
```

**Comment by badboy** [comment]
circleci failed earlier, redoing it went fine: https://app.circleci.com/pipelines/github/mozilla/glean/8759/workflows/48007689-415d-4345-af8d-74c8e84e57ee

**Pull request merged**